platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/projectmodule/DefaultProjectPublicationRegistry.java

```diff
@@ -29,35 +30,36 @@
 import java.util.List;
 
 public class DefaultProjectPublicationRegistry implements ProjectPublicationRegistry, HoldsProjectState {
-    private final SetMultimap<Path, Reference<?>> publicationsByProject = LinkedHashMultimap.create();
+    private final SetMultimap<Path, ProjectPublication> publicationsByProjectId = LinkedHashMultimap.create();
+    private final SetMultimap<BuildIdentifier, PublicationForProject<?>> publicationsByBuildId = LinkedHashMultimap.create();
 
     @Override
-    public <T extends ProjectPublication> Collection<T> getPublications(Class<T> type, Path projectIdentityPath) {
-        synchronized (publicationsByProject) {
-            Collection<Reference<?>> projectPublications = publicationsByProject.get(projectIdentityPath);
+    public <T extends ProjectPublication> Collection<T> getPublicationsForProject(Class<T> type, Path projectIdentityPath) {
+        synchronized (publicationsByProjectId) {
+            Collection<ProjectPublication> projectPublications = publicationsByProjectId.get(projectIdentityPath);
             if (projectPublications.isEmpty()) {
                 return Collections.emptyList();
             }
             List<T> result = new ArrayList<>(projectPublications.size());
-            for (Reference<?> reference : projectPublications) {
-                if (type.isInstance(reference.get())) {
-                    result.add(type.cast(reference.get()));
+            for (ProjectPublication publication : projectPublications) {
+                if (type.isInstance(publication)) {
+                    result.add(type.cast(publication));
                 }
             }
             return result;
         }
     }
 
     @Override
-    public <T extends ProjectPublication> Collection<Reference<T>> getPublications(Class<T> type) {
-        synchronized (publicationsByProject) {
-            Collection<Reference<?>> allPublications = publicationsByProject.values();
-            if (allPublications.isEmpty()) {
+    public <T extends ProjectPublication> Collection<PublicationForProject<T>> getPublicationsForBuild(Class<T> type, BuildIdentifier buildIdentity) {
```

Comment:
The significant changes here are that this service was made to be build tree scoped instead of build scoped. This way, we can ask for the publications of any project, regardless of build, allowing us to detect the "published" coordinates of any project in the build tree, not just projects in our build.